math/matrix/submatrix.go

```diff
@@ -0,0 +1,64 @@
+package matrix
+
+import (
+	"context"
+	"errors"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// SubMatrix extracts a submatrix from the current matrix.
+func (m Matrix[T]) SubMatrix(rowStart, colStart, numRows, numCols int) (Matrix[T], error) {
+	if rowStart < 0 || colStart < 0 || numRows < 0 || numCols < 0 {
+		return Matrix[T]{}, errors.New("negative dimensions are not allowed")
+	}
+
+	if rowStart+numRows > m.rows || colStart+numCols > m.columns {
+		return Matrix[T]{}, errors.New("submatrix dimensions exceed matrix bounds")
+	}
+
+	var zeroVal T
+	if numRows == 0 || numCols == 0 {
+		return New(numRows, numCols, zeroVal), nil // Return an empty matrix
+	}
+
+	subMatrix := New(numRows, numCols, zeroVal)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel() // Make sure it's called to release resources even if no errors
+
+	var eg errgroup.Group
+	var firstErr error
+
+	for i := 0; i < numRows; i++ {
+		i := i // Capture the loop variable for the goroutine
```

Comment:
😂 This has caused so many bugs for me before 😂

Fortunately there is already a kinda fix in go1.21 - [ref](https://github.com/golang/go/wiki/LoopvarExperiment) to the experiment which is tracking the issue.

Reply:
@tjgurwara99 , According to the references, i think we have to use line 34 or run `GOEXPERIMENT=loopvar go test` instead of `go test`? Or there is any other possible solution which i didn't get? as the change need to be made is not yet done and is expected to be done in go1.22 or later (which I have understood).

Reply:
Oh sorry, I think I miscommunicated here - I didn't ask you to change anything here. Sorry if that confused you